### PR TITLE
feat: RE2 Regex tester

### PR DESCRIPTION
**Pull request by krzychukula**
## About the changes

It's a bunch of copy&paste from the AddValue components/logic with addition for the REGEX. 

I extracted buttons for inverting and case sensitivity to be able to use them in the popover. That turned out to not bee needed 

## Discussion points

I have mixed feelings about all those refs but it works so we plan to have someone test this and get feedback. 

# How it looks

Zrzut ekranu 2026-02-26 o 16 39 07

### CODE REVIEW

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexValueEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+        setInputValue(initialValue || '');
+        setError(validateValue(initialValue || ''));
+    }, [editingOpen, initialValue]);
+
+    const handleOnEnterInRegexInput = () => {
+        if (!inputValue?.trim()) {
+            setError('Value cannot be empty or whitespace');
+            return;
+        }
+        // otherwise move to the test input
```
Comment (nunogois) [inline frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexValueEditor.tsx]: We usually prefer no inline comments, unless they're **really** needed. I think most of these can be understood from looking at the code.

File: biome.json [hunk 2]
```diff
@@ -32,7 +32,15 @@
         "noUnusedTemplateLiteral": "off",
         "useSingleVarDeclarator": "off",
         "noUselessElse": "off",
-        "useNodejsImportProtocol": "off"
+        "useNodejsImportProtocol": "off",
+        "noRestrictedGlobals": {
+          "level": "error",
+          "options": {
+            "deniedGlobals": {
+              "open": "Did you mean a local variable? `open` is `window.open`. Declare a local variable instead."
```
Comment (krzychukula) [inline biome.json]: This is needed because we have `"useExhaustiveDependencies": "off",`. Unfortunately enabling that is not that easy so I moved that to a separate issues and this should be a small help:  https://linear.app/unleash/issue/2-4242/enable-biome-useexhaustivedependencies-to-catch-react-bugs

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/ToggleConstraintCaseSensitivity.tsx [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import { styled } from '@mui/material';
```
Comment (krzychukula) [inline frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/ToggleConstraintCaseSensitivity.tsx:1]: This is moved from EditableContraint.tsx

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/ToggleConstraintInverted.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
+import { styled } from '@mui/material';
```
Comment (krzychukula) [inline frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/ToggleConstraintInverted.tsx:1]: Extract component from EditableContraint.tsx

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/EditableConstraint.tsx [hunk 5]
```diff
@@ -228,6 +188,77 @@ const TopRowInput: FC<{
...
+                validator={validator}
+                editingOpen={editingOpen}
+                helpText={
+                    <>
+                        A regex value should be a valid{' '}
+                        <Link
+                            href='https://github.com/google/re2/wiki/Syntax'
+                            target='_blank'
+                            rel='noopener noreferrer'
+                            sx={{
```
Comment (krzychukula) [inline frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/EditableConstraint.tsx]: extract StyledLink here.

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexConstraintValueWidget.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+type Props = {
+    removeValue: () => void;
+    currentValue?: string;
+    editingOpen: boolean;
+    setEditingOpen: (open: boolean) => void;
+    validator: (value: string) => ConstraintValidatorOutput;
+};
+
+export const AddRegexConstraintValueWidget: FC<Props> = ({
```
Comment (krzychukula) [inline frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexConstraintValueWidget.tsx:35]: Consider renaming "Widget" to "Chip" in the follow-up PR.

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexValueEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+    fontSize: 'inherit',
+}));
+
+const StyledNoMatchIcon = styled(HighlightOff)(({ theme }) => ({
+    color: theme.palette.warning.main,
+    fontSize: 'inherit',
+}));
+
+const MatchIndicator: FC<{ passes: boolean; testString: string }> = ({
+    passes,
```
Comment (krzychukula) [inline frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexValueEditor.tsx]: rename to `match`

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexValueEditor.tsx [hunk 8]
```diff
@@ -0,0 +1,522 @@
...
+    useEffect(() => {
+        setRegexTestInputs((prev) =>
+            prev.map((input) => ({
+                ...input,
+                match: matchesRegex(input.testString),
+            })),
+        );
+    }, [matchesRegex]);
+
+    useEffect(() => {
```
Comment (krzychukula) [inline frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexValueEditor.tsx]: Group into chunks.

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexValueEditor.tsx [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+                <MatchIndicator
+                    match={input.match}
+                    testString={input.testString}
+                />
+            </StyledListItem>
+        );
+    },
+);
+
+const RegexTestValues: FC<{
```
Comment (krzychukula) [inline frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexValueEditor.tsx:200]: I could move this to a separate file as well, but it's not really reusable as it need so be shown in really specific context to make sense.

File: biome.json (hunk 2 above)
Comment (kwasniew) [reply biome.json]: which code in this PR triggered this change?

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/ToggleConstraintInverted.tsx [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+            <StyledButton
+                type='button'
+                onClick={onToggleInverted}
+                aria-label={
+                    inverted
+                        ? 'The constraint operator is exclusive.'
+                        : 'The constraint operator is inclusive.'
+                }
+            >
+                <ConstrainInversionIcon inverted={!!inverted} />
```
Comment (kwasniew) [inline frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/ToggleConstraintInverted.tsx]: Boolean(inverted) is more idiomatic in this codebase

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/ToggleConstraintInverted.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    );
+};
+
+interface ToggleInvertedProps {
+    onToggleInverted: () => void;
+    inverted?: boolean;
+}
+
+export const ToggleConstraintInverted = ({
+    inverted,
```
Comment (kwasniew) [inline frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/ToggleConstraintInverted.tsx:51]: can we have a default value/param

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/ToggleConstraintInverted.test.tsx [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    test('defaults to inclusive behavior when inverted is undefined', () => {
+        render(<ToggleConstraintInverted onToggleInverted={vi.fn()} />);
+        expect(
+            screen.getByRole('button', {
+                name: 'The constraint operator is inclusive.',
+            }),
+        ).toBeInTheDocument();
+    });
+
+    test('calls onToggleInverted when clicked', () => {
```
Comment (kwasniew) [inline frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/ToggleConstraintInverted.test.tsx]: I don't see much value in tests verifying mock was called. Don't think it will prevent us from actual errors

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/EditableConstraint.tsx [hunk 13]
```diff
@@ -341,26 +371,14 @@ export const EditableConstraint: FC<Props> = ({
...
-                                        {localConstraint.caseInsensitive ? (
-                                            <StyledCaseInsensitiveIcon aria-label='The match is not case sensitive.' />
-                                        ) : (
-                                            <StyledCaseSensitiveIcon aria-label='The match is case sensitive.' />
-                                        )}
-                                    </StyledButton>
-                                </HtmlTooltip>
+                                <ToggleConstraintCaseSensitivity
+                                    caseInsensitive={
+                                        !!localConstraint.caseInsensitive
```
Comment (kwasniew) [inline frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/EditableConstraint.tsx]: Boolean()

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexConstraintValueWidget.test.tsx [hunk 14]
```diff
@@ -0,0 +1,170 @@
+import { screen, fireEvent } from '@testing-library/react';
```
Comment (kwasniew) [inline frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexConstraintValueWidget.test.tsx:1]: a general comment about all the tests in this PR. I'd remove all of them. Instead I'd create one or two high level tests that simulate e2e what you showed me in the pair review. Testing the entire regex form. With jsdom/vitest. I've never seen any of those localized component tests catching anything useful.

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/ToggleConstraintInverted.tsx (hunk 11 above)
Comment (krzychukula) [reply frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/ToggleConstraintInverted.tsx:51]: I would rather always force caller to pass it. It doesn't have an inherent default value.

File: frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexConstraintValueWidget.test.tsx (hunk 14 above)
Comment (krzychukula) [reply frontend/src/component/feature/FeatureStrategy/FeatureStrategyConstraints/EditableConstraint/AddRegexConstraintValueWidget.test.tsx:1]: Not sure. I would need to remove some of the tests for corner-cases and just create the same tests again for the common path. 

I removed the mock tests from the other buttons, but I would rather leave this one 🤔
